src/posix/platform/netif.cpp

```diff
@@ -298,36 +325,81 @@ static uint8_t NetmaskToPrefixLength(const struct sockaddr_in6 *netmask)
 
-    otError error = OT_ERROR_NONE;
+    struct
+    {
+        struct nlmsghdr  nh;
+        struct ifaddrmsg ifa;
+        char             buf[512];
+    } req;
 
-    assert(sInstance == aInstance);
+    memset(&req, 0, sizeof(req));
 
-    VerifyOrExit(sIpFd >= 0, error = OT_ERROR_INVALID_STATE);
+    req.nh.nlmsg_len   = NLMSG_LENGTH(sizeof(struct ifaddrmsg));
+    req.nh.nlmsg_flags = NLM_F_REQUEST | NLM_F_ACK | NLM_F_CREATE | NLM_F_EXCL;
+    req.nh.nlmsg_type  = aIsAdded ? RTM_NEWADDR : RTM_DELADDR;
+    req.nh.nlmsg_pid   = 0;
+    req.nh.nlmsg_seq   = ++sNetlinkSequence;
 
-#if defined(__linux__)
+    req.ifa.ifa_family    = AF_INET6;
+    req.ifa.ifa_prefixlen = aAddressInfo.mPrefixLength;
+    req.ifa.ifa_flags     = aAddressInfo.mIsAnycast ? IFA_F_NODAD : 0;
+    req.ifa.ifa_scope     = aAddressInfo.mScope;
+    req.ifa.ifa_index     = gNetifIndex;
+
+    rta           = reinterpret_cast<struct rtattr *>((reinterpret_cast<char *>(&req)) + NLMSG_ALIGN(req.nh.nlmsg_len));
+    rta->rta_type = IFA_LOCAL;
+    rta->rta_len  = RTA_LENGTH(sizeof(*aAddressInfo.mAddress));
+
+    memcpy(RTA_DATA(rta), aAddressInfo.mAddress, sizeof(*aAddressInfo.mAddress));
+
+    req.nh.nlmsg_len = NLMSG_ALIGN(req.nh.nlmsg_len) + rta->rta_len;
+
+    if (aAddressInfo.mIsAnycast)
     {
-        struct in6_ifreq ifr6;
-        memcpy(&ifr6.ifr6_addr, &aAddress, sizeof(ifr6.ifr6_addr));
+        struct ifa_cacheinfo cacheinfo;
 
-        ifr6.ifr6_ifindex   = static_cast<int>(gNetifIndex);
-        ifr6.ifr6_prefixlen = aPrefixLength;
+        rta           = reinterpret_cast<struct rtattr *>((reinterpret_cast<char *>(rta)) + rta->rta_len);
+        rta->rta_type = IFA_CACHEINFO;
+        rta->rta_len  = RTA_LENGTH(sizeof(cacheinfo));
 
-        if (aIsAdded)
-        {
-            VerifyOrDie(ioctl(sIpFd, SIOCSIFADDR, &ifr6) == 0, OT_EXIT_ERROR_ERRNO);
-        }
-        else
-        {
-            VerifyOrExit(ioctl(sIpFd, SIOCDIFADDR, &ifr6) == 0, perror("ioctl"); error = OT_ERROR_FAILED);
-        }
+        memset(&cacheinfo, 0, sizeof(cacheinfo));
+        cacheinfo.ifa_valid = UINT32_MAX;
+
+        memcpy(RTA_DATA(rta), &cacheinfo, sizeof(cacheinfo));
+
+        req.nh.nlmsg_len += rta->rta_len;
+    }
+
+    if (send(sNetlinkFd, &req, req.nh.nlmsg_len, 0) != -1)
+    {
+        otLogInfoPlat("Sent request#%u to %s %s/%u", sNetlinkSequence, (aIsAdded ? "add" : "remove"),
+                      Ip6AddressString(aAddressInfo.mAddress).Get(), aAddressInfo.mPrefixLength);
     }
+    else
+    {
+        otLogInfoPlat("Failed to send request#%u to %s %s/%u", sNetlinkSequence, (aIsAdded ? "add" : "remove"),
+                      Ip6AddressString(aAddressInfo.mAddress).Get(), aAddressInfo.mPrefixLength);
+    }
+}
+#endif // defined(__linux__)
+
+static void UpdateUnicast(otInstance *aInstance, const otIp6AddressInfo &aAddressInfo, bool aIsAdded)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+
+    assert(sInstance == aInstance);
+    assert(sIpFd >= 0);
```

Comment:
Should we use `OT_ASSERT` instead?

Reply:
There are multiple instances of assert() in posix platform. I tend to address this issue together in a separate PR. Thoughts?

Reply:
`OT_ASSERT` is defined in OT core. We can't use it in platform code.